https://stackoverflow.com/questions/54754153/autograd-grad-for-tensor-in-pytorch

In [8]:
import torch
from torch.autograd import grad
import torch.nn as nn

In [9]:
# Create some dummy data.
x = torch.ones(2, 2, requires_grad=True)
gt = torch.ones_like(x, requires_grad=True) * 16 - 0.5 # ground-truths

x, gt

(tensor([[1., 1.],
         [1., 1.]], requires_grad=True),
 tensor([[15.5000, 15.5000],
         [15.5000, 15.5000]], grad_fn=<SubBackward0>))

In [10]:
# We will use MSELoss as an examples.
loss_fn = nn.MSELoss()

In [11]:
# Do some computations.
v = x + 2
y = v ** 2

v.retain_grad()
y.retain_grad()

In [12]:
# Compute loss
loss = loss_fn(y, gt)

print(f"Loss: {loss}")

Loss: 42.25


In [6]:
# Now compute gradients:
d_loss_dx = grad(outputs=loss, inputs=x)
print(f"dloss/dx:\n {d_loss_dx}")

dloss/dx:
 (tensor([[-19.5000, -19.5000],
        [-19.5000, -19.5000]]),)


In [13]:
# Now compute gradients:
d_loss_dy = grad(outputs=loss, inputs=y)
print(f"dloss/dy:\n {d_loss_dy}")

dloss/dy:
 (tensor([[-3.2500, -3.2500],
        [-3.2500, -3.2500]]),)


In [35]:
y.grad

tensor([[-3.2500, -3.2500],
        [-3.2500, -3.2500]])

In [34]:
d_loss_dx[0] / y.grad

tensor([[6., 6.],
        [6., 6.]])

$$\cfrac{dl}{dy}\cfrac{dy}{dx}=-19.5$$
$$\cfrac{dl}{dy}=-3.25$$
$$\cfrac{dy}{dx}=6$$

In [14]:
v = x + 2
y = v ** 2

try:
    dy_hat_dx = grad(outputs=y, inputs=x)
except RuntimeError as err:
    print(err)

grad can be implicitly created only for scalar outputs


In [20]:
v = x + 2
y = v ** 2

dy_dx = grad(outputs=y, inputs=x, grad_outputs=torch.ones_like(y))
print(f'dy/dx:\n {dy_dx}')

dv_dx = grad(outputs=v, inputs=x, grad_outputs=torch.ones_like(v))
print(f'dv/dx:\n {dv_dx}')

dy/dx:
 (tensor([[6., 6.],
        [6., 6.]]),)
dv/dx:
 (tensor([[1., 1.],
        [1., 1.]]),)
